### PR TITLE
support yarn in turbomind backend

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

support yarn in torbumind backend

https://github.com/huggingface/transformers/blob/f2c388e3f946862f657acc1e21b272ec946fc66c/src/transformers/modeling_rope_utils.py#L163

### CODE REVIEW

**Comment by lvhan028** [comment]
@irexyc may check the pytorch engine as well. Please support it if pytorch engine has no such feature

File: lmdeploy/pytorch/nn/rotary_embedding.py [hunk 1]
```diff
@@ -1,12 +1,91 @@
...
+    rope_scaling = config.rope_scaling
+    scaling_factor = rope_scaling['factor']
+    return dict(emb_type=RopeType.DynamicNTKScaling,
+                scaling_factor=scaling_factor)
+
+
+def _get_yarn_parameters(config: PretrainedConfig):
+    """get yarn parameters."""
+    rope_scaling = config.rope_scaling
+    scaling_factor = rope_scaling['factor']
```
Comment (grimoire) [inline lmdeploy/pytorch/nn/rotary_embedding.py:34]: This is not general enough, some model might name it 'scaling_factor' or something else.

File: src/turbomind/kernels/attention/rotary_embedding.h [hunk 2]
```diff
@@ -110,6 +116,15 @@ struct FastRoPE {
                 inv_freq_[i / 2] = (1 - smooth) * freq * llama3_inv_scaling_factor + smooth * freq;
             }
         }
+        if (yarn_ramp_max) {
+            PRAGMA_UNROLL
+            for (int i = 0; i < N; i += 2) {
+                auto  freq       = inv_freq_[i / 2];
+                float alpha      = ((idx + i) / 2 - yarn_ramp_min) / (yarn_ramp_max - yarn_ramp_min);
+                alpha            = fmaxf(0.f, fminf(1.f, alpha));
+                inv_freq_[i / 2] = freq * (1 - alpha + alpha / factor);
```
Comment (lzhangzz) [inline src/turbomind/kernels/attention/rotary_embedding.h:125]: All these expensive divisions can be done in host code. See how the llama3 rope is implemented just a few lines above.

File: lmdeploy/pytorch/nn/rotary_embedding.py (hunk 1 above)
Comment (irexyc) [reply lmdeploy/pytorch/nn/rotary_embedding.py:34]: I followed [transformers](https://github.com/huggingface/transformers/blob/5f4ee98a7ade33e1c54fdd6181d04ee7b426b392/src/transformers/modeling_rope_utils.py#L403)(ROPE_INIT_FUNCTIONS) which should suit for builtin models.

Maybe custom models like [deepseek](https://github.com/InternLM/lmdeploy/blob/main/lmdeploy/pytorch/models/deepseek_v2.py#L548) should not use `build_rotary_params` and build rotary_params in a custom way.

File: lmdeploy/turbomind/deploy/config.py [hunk 3]
```diff
@@ -53,13 +53,16 @@ class ModelConfig:
 class AttentionConfig:
     rotary_embedding: int = 128
     rope_theta: float = 10000.0
+    attention_factor: float = None
```
Comment (lvhan028) [inline lmdeploy/turbomind/deploy/config.py:56]: https://github.com/huggingface/transformers/blob/f2c388e3f946862f657acc1e21b272ec946fc66c/src/transformers/modeling_rope_utils.py#L189
There is another parameter "partial_rotary_factor"

File: lmdeploy/turbomind/deploy/source_model/llama.py [hunk 4]
```diff
@@ -236,6 +239,10 @@ def model_info(self):
                     else llama3_scaling_type
                 if scaling_type == 'dynamic':
                     use_dynamic_ntk = 1
+                attention_factor = model_arg['rope_scaling'].get(
+                    'attention_factor', None)
```
Comment (lvhan028) [inline lmdeploy/turbomind/deploy/source_model/llama.py:243]: https://github.com/huggingface/transformers/blob/f2c388e3f946862f657acc1e21b272ec946fc66c/src/transformers/modeling_rope_utils.py#L198
```
attention_factor = config.rope_scaling.get("attention_factor")
    if attention_factor is None:
        attention_factor = 0.1 * math.log(factor) + 1.0
```

**Review by grimoire — APPROVED** [review]

File: lmdeploy/turbomind/deploy/source_model/llama.py (hunk 4 above)
Comment (irexyc) [reply lmdeploy/turbomind/deploy/source_model/llama.py]: The [previous code](https://github.com/InternLM/lmdeploy/blob/152d4be19bf271a8a47ec5c54c8a3da7a3dc651c/src/turbomind/models/llama/unified_attention_layer.cc#L333-L338) had this logic. And add this when converting the model as well.

File: lmdeploy/turbomind/deploy/config.py (hunk 3 above)
Comment (irexyc) [reply lmdeploy/turbomind/deploy/config.py:56]: It seems the `partial_rotary_factor` changes the dim of `default`, `dynamic ntk` and `yarn`

**Review by lzhangzz — APPROVED** [review]

**Pull request merged**